lib/progress-state.coffee

```diff
@@ -0,0 +1,132 @@
+###*
+# @module resin/progress-state
+# @protected
+###
+
+_ = require('lodash')
+
+###*
+# Represent a progress state
+# @name ProgressState
+# @protected
+# @class
+# @param {Object} options - progress state options
+#
+# A ProgressState instance consists of the following fields:
+#
+# - total.
+# - percentage.
+# - eta.
+# - received.
+# - delta.
+#
+# @throws {Error} If the state options are invalid
+###
+class ProgressState
+
+	constructor: (options) ->
+
+		# Total option
+
+		if options.total?
+
+			if not _.isNumber(options.total) or options.total < 0
+				throw new Error("Invalid total option: #{options.total}")
+
+		# Percentage option
+
+		if options.percentage?
+
+			if not _.isNumber(options.percentage) or options.percentage < 0 or options.percentage > 100
+				throw new Error("Invalid percentage option: #{options.percentage}")
+
+		# ETA option
+
+		if options.eta?
+
+			if not _.isNumber(options.eta) or options.eta < 0
+				throw new Error("Invalid eta option: #{options.eta}")
+
+		# Received option
+
+		if not options.received?
+			throw new Error('Missing received option')
+
+		if not _.isNumber(options.received) or options.received < 0
+			throw new Error("Invalid received option: #{options.received}")
+
+		if options.total? and options.received > options.total
+			throw new Error("Received option can't be higher than total: #{options.received} > #{options.total}")
+
+		# Delta option
+
+		if not options.delta?
+			throw new Error('Missing delta option')
+
+		if not _.isNumber(options.delta) or options.delta < 0
+			throw new Error("Invalid delta option: #{options.delta}")
+
+		if options.total? and options.delta > options.total
+			throw new Error("Delta option can't be higher than total: #{options.delta} > #{options.total}")
+
+		if options.delta > options.received
+			throw new Error("Delta option can't be higher than received: #{options.delta} > #{options.received}")
+
+		_.extend(this, options)
+
+getCurrentTime = ->
+	return new Date().getTime()
+
+###*
+# createFromNodeRequestProgress callback
+# @callback module:resin/auth~createFromNodeRequestProgressCallback
+# @param {module:resin/progress-state~ProgressState} state - progress state
+###
+
+###*
+# @summary Create a ProgressState listener from a node-request-progress state
+# @protected
+# @function
+#
+# @param {module:resin/auth~createFromNodeRequestProgressCallback} callback - callback
+#
+# @example
+# request = require('request')
+# progress = require('request-progress')
+#
+# onProgressCallback = (state) ->
+#		console.log(state.percentage)
+#
+# progress(request(options))
+#		.on('progress', ProgressState.createFromNodeRequestProgress(onProgressCallback))
+#		.on('error', ...)
+#		.on('close', ...)
+###
+ProgressState.createFromNodeRequestProgress = (callback) ->
+	time = getCurrentTime()
+	received = 0
+
+	return (state) ->
+		newTime = getCurrentTime()
```

Comment:
It indeed increases readability. Fixed in https://github.com/resin-io/resin-sdk/commit/16d6049fdad45314ae7579b3eb054d644e134c30.